### PR TITLE
Introduce state computer protocol updates

### PR DESCRIPTION
**Pull request by LukasGasior1**
To be added in a follow up PR:
- protocol update execution
- tests for protocol updates \w state version bounds

### CODE REVIEW

File: core-rust/state-manager/src/protocol/protocol_config.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub enum ProtocolUpdateEnactmentCondition {
+    EnactWhenSupportedAndWithinBounds {
+        lower_bound: ProtocolUpdateEnactmentBound,
+        upper_bound: ProtocolUpdateEnactmentBound,
+        support_type: ProtocolUpdateSupportType,
+    },
+    EnactUnconditionallyAtEpoch(Epoch),
+    EnactUnconditionallyAtStateVersion(StateVersion),
```
Comment (jakrawcz-rdx) [inline core-rust/state-manager/src/protocol/protocol_config.rs:50]: just as discussed on our call: it still seems to me like some portion of conceptual complexity (and code) could be gone if we discarded the "unconditional" methods (and used `required_ratio_of_stake_supported: 0.0` for "asap updates"; and stored the "already-enacted-by-quorum" updates in a map). Please see if there are any true blockers to migrate to that approach.

File: core-rust-bridge/src/main/java/com/radixdlt/protocol/ProtocolUpdateEnactmentBound.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  static void registerCodec(CodecMap codecMap) {
+    codecMap.register(
+        ProtocolUpdateEnactmentBound.class,
+        codecs ->
+            EnumCodec.fromPermittedRecordSubclasses(ProtocolUpdateEnactmentBound.class, codecs));
+  }
+
+  record Epoch(UInt64 epoch) implements ProtocolUpdateEnactmentBound {}
+
+  record StateVersion(UInt64 stateVersion) implements ProtocolUpdateEnactmentBound {}
```
[the remark this answers is not in the document]
Comment (jakrawcz-rdx) [reply core-rust-bridge/src/main/java/com/radixdlt/protocol/ProtocolUpdateEnactmentBound.java:81]: I am also a fan of leaving the state versions out.

Possibly there is a way of avoiding them completely (i.e. only ever switch protocol on epoch boundaries) by introducing a way of doing `Engine.forceEpochChangeRightNow()`...?

File: core-rust/state-manager/src/store/rocks_db.rs [hunk 3]
```diff
@@ -206,6 +207,16 @@ impl VersionedCf<Epoch, LedgerProof> for EpochLedgerProofsCf {
     type VersionedValue = VersionedLedgerProof;
 }
 
+/// Ledger proofs of protocol updates.
+/// Schema: `Epoch.to_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
+/// Note: This duplicates a small subset of [`StateVersionToLedgerProof`]'s values.
```
Comment (jakrawcz-rdx) [inline core-rust/state-manager/src/store/rocks_db.rs:212]:
```suggestion
/// Note: This duplicates a small subset of [`LedgerProofsCf`]'s values.
```

_(plz also propagate the fix to my line that you copied :p)_

File: core-rust/state-manager/src/store/rocks_db.rs [hunk 4]
```diff
@@ -206,6 +207,16 @@ impl VersionedCf<Epoch, LedgerProof> for EpochLedgerProofsCf {
     type VersionedValue = VersionedLedgerProof;
 }
 
+/// Ledger proofs of protocol updates.
+/// Schema: `Epoch.to_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
```
Comment (jakrawcz-rdx) [inline core-rust/state-manager/src/store/rocks_db.rs:211]:
```suggestion
/// Schema: `StateVersion.to_be_bytes()` -> `scrypto_encode(VersionedLedgerProof)`
```

File: core-rust-bridge/src/main/java/com/radixdlt/protocol/ProtocolUpdateEnactmentBound.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (LukasGasior1) [reply core-rust-bridge/src/main/java/com/radixdlt/protocol/ProtocolUpdateEnactmentBound.java:81]: I removed state version updates for now in a follow up PR: https://github.com/radixdlt/babylon-node/pull/790

**Comment by LukasGasior1** [comment]
@talekhinezh @jakrawcz-rdx I'm going to close this PR in favor of https://github.com/radixdlt/babylon-node/pull/790 , which includes the whole feature.
@jakrawcz-rdx I've addressed your comment-related comments there. Let's also discuss unconditional vs 0.0 required support updates there.

**Pull request closed without merging**